File: lib/decidim/spid.rb
# Copyright (C) 2022 Formez PA
# This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, version 3.
# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
# You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>

# frozen_string_literal: true

require 'deface'
require "omniauth/strategies/spid_saml"

require_relative "spid/version"
require_relative "spid/engine"
require_relative "spid/admin"
require_relative "spid/admin_engine"
require_relative "spid/authentication"
require_relative "spid/verification"
require_relative "spid/component"

module Decidim
  module Spid
    autoload :Tenant, "decidim/spid/tenant"

    class << self
      def tenants
        @tenants ||= []
      end

      def test!
        @test = true
      end

      def configure(&block)
        tenant = Decidim::Spid::Tenant.new(&block)
        tenants.each do |existing|
          if tenant.name == existing.name
            raise(
              TenantSameName,
              "Definisci il nome del Tenant. Il nome \"#{tenant.name}\" è già in uso."
            )
          end

          match = tenant.name =~ /^#{existing.name}/
          match ||= existing.name =~ /^#{tenant.name}/
          next unless match

        end

        tenants << tenant
      end

      def setup!
        raise "Il modulo SPID è già stato inizializzato!" if initialized?

        @initialized = true
        tenants.each(&:setup!)
      end

      def find_tenant(name)
        Decidim::Spid.tenants.select { |a| a.name == name}.try(:first)
      end

      private

      def initialized?
        @initialized
      end
    end

    class TenantSameName < StandardError; end

    class InvalidTenantName < StandardError; end
  end
end

File: app/packs/src/decidim/spid/spid.js
// Copyright (C) 2022 Formez PA
// This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, version 3.
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
// You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>

'use strict';

var _popup = function _popup() {
    var container = document.querySelector('.spid-selector-container');
    // Se il container principale non esiste, esci subito dalla funzione
    if (!container) return;

    var button = container.querySelector('.italia-it-button');
    var popupContainer = container.querySelector('.spid-container');
    var questionContainer = container.querySelector('.spid-alert');
    var questionSelector = container.querySelector('.spid-question-mark');

    // Funzione di chiusura globale sicura
    document.onclick = function(e) {
        if (popupContainer) popupContainer.classList.remove('opened');
        if (questionContainer) questionContainer.classList.remove('opened');
    };

    // Gestione bottone SPID
    if (button && popupContainer && questionContainer) {
        button.onclick = function(e) {
            e.stopPropagation();
            questionContainer.classList.remove('opened');
            var list = popupContainer.querySelectorAll('div[data-idp]');
            for (var i = list.length; i >= 0; i--) {
                popupContainer.prepend(list[Math.random() * i | 0]);
            }
            popupContainer.classList.toggle('opened');
        };
    }

    // Gestione alert info
    if (questionContainer) {
        questionContainer.onclick = function(e) { e.stopPropagation(); };
    }
    
    if (questionSelector && popupContainer && questionContainer) {
        questionSelector.onclick = function(e) {
            e.stopPropagation();
            popupContainer.classList.remove('opened');
            questionContainer.classList.toggle('opened');
        };
    }

    // Gestione click sugli IDP
    if (popupContainer) {
        popupContainer.querySelectorAll("[data-idp]").forEach(function (btn) {
            btn.addEventListener('click', function (e) {
                e.stopPropagation();
                var form = btn.querySelector("form");
                if (form) form.submit();
            });
        });
    }
};

document.addEventListener("DOMContentLoaded", _popup);
document.addEventListener('turbolinks:load', _popup);